cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1807,7 +1808,8 @@ struct CaseWhenFunctor<Type, enable_if_base_binary<Type>> {
 };
 
 template <typename Type>
-struct CaseWhenFunctor<Type, enable_if_var_size_list<Type>> {
+struct CaseWhenFunctor<
+    Type, enable_if_t<is_base_list_type<Type>::value || is_list_view_type<Type>::value>> {
```

Comment:
The list-view types should have their own specialization because they have a super-power that classic list types don't have: you can append the child values in any order and adjust the offset/size pairs of any random position to point to that area.

Feel free to copy and paste this class and enable_if it for `is_list_view` with a `TODO(GH-<issue number>): a more efficient implementation for list-views is possible` comment mentioned an issue you can create about this. Then you can tackle the optimization in a separate PR.

Reply:
> you can append the child values in any order and adjust the offset/size pairs of any random position to point to that area.

Thanks for your suggestion, i will dig into the list-view types.

> Feel free to copy and paste this class and enable_if it for `is_list_view` with a `TODO(GH-<issue number>): a more efficient implementation for list-views is possible` comment mentioning an issue you can create about this. Then we can tackle the optimization in a separate PR.

Agree!

Reply:
1. New issue for optimization list-view types: https://github.com/apache/arrow/issues/41453
2. New commit include a list-view type's benchmark for our performance regression in the future